### PR TITLE
feat(firestore-bigquery-change-tracker): add TimePartitioning on selected Firstore Document field

### PR DESCRIPTION
**Pull request by posiek07**
Needs to be published on NPM before: #863 

## Feature
 - add partitioning on selected top-layer Firestore Document field (Firestore Timestamp or DATE, DATETIME, TIMESTAMP BigQuery format) that will be imported to schema and Table will be TimePartitioned on.
  
 ## Jobs
- [x] add conditional partitioning parameters to `FirestoreBigQueryEventHistoryTracker`
- [x] add type `timePartitioning(changed from table Partitioning), timePartitioningField, timePartitioningFieldType, timePartitioningFirestoreField` to FirestoreBigQueryEventHistoryTrackerConfig
- [x] add `record` function custom TimePartitioning field property if all TimePartitioning parameters supplied and Firestore Document field exists.
- [x] add `getNewPartitionedField` function for generating partition table or view schema with extra TimePartitioning partitioning field/column
- [x] add error handling logs for table inserts
- [x] make manual tests screenshots with update on big-query-export extension
 
# Tests
## Update Engine
Screenshot 2022-01-19 at 13 42 53
Screenshot 2022-01-19 at 13 41 31

## Test for Firestore Timestamp or BigQuery TIMESTAMP field partitioning
Screenshot 2022-01-19 at 13 41 20
Screenshot 2022-01-19 at 13 46 18

**Firestore Timestamp results**
Screenshot 2022-01-19 at 13 41 48
Screenshot 2022-01-19 at 13 42 07
Screenshot 2022-01-19 at 13 46 49

**BigQuery TIMESTAMP result**
Screenshot 2022-01-19 at 14 36 26
Screenshot 2022-01-19 at 12 49 43
Screenshot 2022-01-19 at 14 39 36
Screenshot 2022-01-19 at 14 39 12

## Test BigQuery DATE field partitioning
Screenshot 2022-01-19 at 13 58 09
Screenshot 2022-01-19 at 14 00 23

**BigQuery DATE result**
Screenshot 2022-01-19 at 13 58 57
Screenshot 2022-01-19 at 13 57 46
Screenshot 2022-01-19 at 14 03 34

## Test BigQuery DATETIME field partitioning
Screenshot 2022-01-19 at 14 24 49
Screenshot 2022-01-19 at 14 26 56

**BigQuery DATETIME result**
Screenshot 2022-01-19 at 14 25 33
Screenshot 2022-01-19 at 14 25 14
Screenshot 2022-01-19 at 14 28 17

## OLD settings test (no new optional fields)
Screenshot 2022-01-19 at 18 03 32
Screenshot 2022-01-19 at 18 03 06

**OLD settings result**
Screenshot 2022-01-19 at 18 06 30
Screenshot 2022-01-19 at 18 05 39
Screenshot 2022-01-19 at 18 04 38

### CODE REVIEW

File: firestore-bigquery-export/firestore-bigquery-change-tracker/src/bigquery/index.ts [hunk 1]
```diff
@@ -68,8 +72,15 @@ export class FirestoreBigQueryEventHistoryTracker
 
   async record(events: FirestoreDocumentChangeEvent[]) {
     await this.initialize();
-
+    const timePartitioningFirestoreField = this.config
+      .timePartitioningFirestoreField;
+    const timePartitioningField = this.config.timePartitioningField;
     const rows = events.map((event) => {
+      const validateIsTimePartitionFirestoreField =
```
Comment (dackers86) [inline firestore-bigquery-export/firestore-bigquery-change-tracker/src/bigquery/index.ts:79]: Could potentially extract this into a private function

File: firestore-bigquery-export/firestore-bigquery-change-tracker/src/bigquery/index.ts [hunk 2]
```diff
@@ -79,6 +90,14 @@ export class FirestoreBigQueryEventHistoryTracker
           document_id: event.documentId,
           operation: ChangeType[event.operation],
           data: JSON.stringify(this.serializeData(event.data)),
+
+          //attach partitioning property if Firestore Field avalible.
+          ...(validateIsTimePartitionFirestoreField && {
```
Comment (dackers86) [inline firestore-bigquery-export/firestore-bigquery-change-tracker/src/bigquery/index.ts:95]: Could also potentially extract into a separate function

File: firestore-bigquery-export/firestore-bigquery-change-tracker/src/bigquery/schema.ts [hunk 3]
```diff
@@ -149,3 +150,31 @@ export const RawChangelogSchema: any = {
...
+    RawChangelogViewSchema.fields.find(
+      (field) => field.name === partitioningField
+    );
+  const newPartitionField = {
+    name: partitioningField,
+    mode: "NULLABLE",
+    type: partitioningFieldType,
+    description: "The document partition field selected by user",
+  };
+  let schema = isView
```
Comment (dackers86) [inline firestore-bigquery-export/firestore-bigquery-change-tracker/src/bigquery/schema.ts:173]: const?

File: firestore-bigquery-export/firestore-bigquery-change-tracker/src/bigquery/schema.ts [hunk 4]
```diff
@@ -70,7 +70,8 @@ export const documentIdField = {
  * We cannot specify a schema for view creation, and all view columns default
  * to the NULLABLE mode.
  */
-export const RawChangelogViewSchema: any = {
+
+export const RawChangelogViewSchema = {
```
Comment (dackers86) [inline firestore-bigquery-export/firestore-bigquery-change-tracker/src/bigquery/schema.ts:74]: Types have been removed, can we add these back in?

File: firestore-bigquery-export/firestore-bigquery-change-tracker/src/bigquery/schema.ts [hunk 5]
```diff
@@ -149,3 +150,31 @@ export const RawChangelogSchema: any = {
...
+  partitioningFieldType: string,
+  isView: boolean
+) => {
+  const isSchemaFieldExist =
+    RawChangelogSchema.fields.find(
+      (field) => field.name === partitioningField
+    ) ||
+    RawChangelogViewSchema.fields.find(
+      (field) => field.name === partitioningField
+    );
```
Comment (dackers86) [inline firestore-bigquery-export/firestore-bigquery-change-tracker/src/bigquery/schema.ts:166]: Can maybe tidy this up or extract out for readability

File: firestore-bigquery-export/firestore-bigquery-change-tracker/src/bigquery/schema.ts (hunk 4 above)
Comment (posiek07) [reply firestore-bigquery-export/firestore-bigquery-change-tracker/src/bigquery/schema.ts:74]: Typescript will automatically assign Types if none are provided. I think it is a better solution than `any` as it is checking them now.

File: firestore-bigquery-export/firestore-bigquery-change-tracker/src/bigquery/schema.ts (hunk 3 above)
Comment (posiek07) [reply firestore-bigquery-export/firestore-bigquery-change-tracker/src/bigquery/schema.ts:173]: I am using the `push` method below so I think need to use `let` in this place

**Review by dackers86 — APPROVED** [review]

**Comment by dackers86** [comment]
Closed via https://github.com/firebase/extensions/pull/891

**Pull request closed without merging**